node/core/pvf/src/prepare/worker.rs

```diff
@@ -314,74 +310,73 @@ async fn recv_request(stream: &mut UnixStream) -> io::Result<(Vec<u8>, PathBuf,
 		loop {
 			let (code, dest, preparation_timeout) = recv_request(&mut stream).await?;
-
 			gum::debug!(
 				target: LOG_TARGET,
 				worker_pid = %std::process::id(),
 				"worker: preparing artifact",
 			);
 
-			// Create a lock flag. We set it when either thread finishes.
-			let lock = Arc::new(AtomicBool::new(false));
+			// Flag used only to signal to the cpu time monitor thread that it can finish.
+			let finished_flag = Arc::new(AtomicBool::new(false));
 			let cpu_time_start = ProcessTime::now();
 
-			// Spawn a new thread that runs the CPU time monitor. Continuously wakes up from
-			// sleeping and then either sleeps for the remaining CPU time, or kills the process if
-			// we exceed the CPU timeout.
-			let (stream_2, cpu_time_start_2, preparation_timeout_2, lock_2) =
-				(stream.clone(), cpu_time_start, preparation_timeout, lock.clone());
-			let handle =
-				thread::Builder::new().name("CPU time monitor".into()).spawn(move || {
-					task::block_on(async {
-						cpu_time_monitor_loop(
-							JobKind::Prepare,
-							stream_2,
-							cpu_time_start_2,
-							preparation_timeout_2,
-							lock_2,
-						)
-						.await;
-					})
-				})?;
-
-			// Prepares the artifact in a separate thread.
-			let result = match prepare_artifact(&code).await {
-				Err(err) => {
-					// Serialized error will be written into the socket.
-					Err(err)
+			// Spawn a new thread that runs the CPU time monitor.
+			let finished_flag_2 = finished_flag.clone();
+			let thread_fut = rt_handle
+				.spawn_blocking(move || {
+					cpu_time_monitor_loop(
+						JobKind::Prepare,
+						cpu_time_start,
+						preparation_timeout,
+						finished_flag_2,
+					)
+				})
+				.fuse();
+			let prepare_fut = rt_handle.spawn_blocking(move || prepare_artifact(&code)).fuse();
+
+			pin_mut!(thread_fut);
+			pin_mut!(prepare_fut);
+
+			let result = select_biased! {
+				// If this future is not selected, the join handle is dropped and the thread will
+				// finish in the background.
+				join_res = thread_fut => {
+					match join_res {
+						Ok(()) => Err(PrepareError::TimedOut),
+						Err(_) => Err(PrepareError::DidNotMakeIt),
+					}
 				},
-				Ok(compiled_artifact) => {
+				compilation_res = prepare_fut => {
 					let cpu_time_elapsed = cpu_time_start.elapsed();
-
-					let lock_result =
-						lock.compare_exchange(false, true, Ordering::Relaxed, Ordering::Relaxed);
-					if lock_result.is_err() {
-						// The other thread is still sending an error response over the socket. Wait on it and
-						// return.
-						let _ = handle.join();
-						// Monitor thread detected timeout and likely already terminated the
-						// process, nothing to do.
-						continue
+					finished_flag.store(true, Ordering::Relaxed);
+
+					match compilation_res.unwrap_or_else(|_| Err(PrepareError::DidNotMakeIt)) {
+						Err(err) => {
+							// Serialized error will be written into the socket.
+							Err(err)
+						},
```

Comment:
Good catch. I think it makes sense to keep the original error string on this error variant, like we do with `ExecuteError::InternalError`. Then we can always log it in the host, and not just here.

Concerning your requested changes though, we can't consolidate the error handling like that because `err` here is a `PrepareError`, whereas above it's a join handle error. The `Err(err) => Err(err)` looks weird, and I've tried to simplify it into a map, but that doesn't work because of the `?` on the `Ok` match arm.